src/intrinsics/ecma262/Object.js

```diff
@@ -60,6 +60,7 @@ export default function(realm: Realm): NativeFunctionValue {
     // 1. Let to be ? ToObject(target).
     let to = To.ToObjectPartial(realm, target);
     let to_must_be_partial = false;
+    let to_can_be_simple = true;
```

Comment:
I don't like that we assume this to be true. Seems dangerous.

Here's my alternative proposal (against master) for how we could express it. Does it make sense?

```diff
@@ -60,6 +60,8 @@ export default function(realm: Realm): NativeFunctionValue {
     // 1. Let to be ? ToObject(target).
     let to = To.ToObjectPartial(realm, target);
     let to_must_be_partial = false;
+    let to_must_be_simple = false;
+    let simple_sources_count = 0;
 
     // 2. If only one argument was passed, return to.
     if (!sources.length) return to;
@@ -81,7 +83,9 @@ export default function(realm: Realm): NativeFunctionValue {
 
         let frm_was_partial = frm.isPartialObject();
         if (frm_was_partial) {
-          if (!frm.isSimpleObject()) {
+          if (frm.isSimpleObject()) {
+            simple_sources_count++
+          } else {
             // If this is not a simple object, it may have getters on it that can
             // mutate any state as a result. We don't yet support this.
             AbstractValue.reportIntrospectionError(nextSource);
@@ -110,9 +114,15 @@ export default function(realm: Realm): NativeFunctionValue {
       if (to_must_be_partial) {
         // Only OK if to is an empty object because nextSource might have
         // properties at runtime that will overwrite current properties in to.
-        // For now, just throw if this happens.
         let to_keys = to.$OwnPropertyKeys();
-        if (to_keys.length !== 0) {
+        if (to_keys.length === 0) {
+          // If all sources are simple, and target has no keys,
+          // we know it is safe to assume target is simple too.
+          if (simple_sources_count === sources.length) {
+            to_must_be_simple = true;
+          }
+        } else {
+          // For now, just throw if this happens.
           AbstractValue.reportIntrospectionError(nextSource);
           throw new FatalError();
         }
@@ -140,7 +150,13 @@ export default function(realm: Realm): NativeFunctionValue {
     }
 
     // 5. Return to.
-    if (to_must_be_partial) to.makePartial();
+    if (to_must_be_partial) {
+      to.makePartial();
+    }
+    if (to_must_be_simple) {
+      to.makeSimple();
+    }
+
     return to;
   });
```

Reply:
I tried that, it fails the my second test as not all sources need to be simple too for the target to be simple.

Reply:
Aah I see what you mean.